its/src/test/java/com/sonar/it/csharp/AutoGeneratedTest.java

```diff
@@ -47,35 +53,96 @@ public void init() {
     }
 
     @Test
-    public void autogenerated_code_reports_no_bugs() throws Exception {
+    public void autogenerated_code_reports_no_bugs_when_generated_not_analyzed() throws Exception {
       final String PROJECT_NAME = "AutoGeneratedFiles";
-      analyzeCoverageTestProject(PROJECT_NAME);
+      analyzeCoverageTestProject(PROJECT_NAME, false);
 
       assertThat(getMeasureAsInt(PROJECT_NAME, "files")).isNull(); // no files are reported to SQ
       assertThat(getMeasureAsInt(PROJECT_NAME, "new_bugs")).isNull();
     }
 
+    @Test
+    public void autogenerated_code_reports_bugs_when_generated_analyzed() throws Exception {
+      final String PROJECT_NAME = "AutoGeneratedFiles";
+      analyzeCoverageTestProject(PROJECT_NAME, true);
+
+      assertThat(getMeasureAsInt(PROJECT_NAME, "ncloc")).isEqualTo(261);
+      assertThat(getMeasureAsInt(PROJECT_NAME, "bugs")).isEqualTo(18);
+    }
+
+    @Test
+    public void non_autogenerated_code_reports_bugs_when_generated_not_analyzed() throws Exception {
+      final String PROJECT_NAME = "NotAutoGeneratedFiles";
+      analyzeCoverageTestProject(PROJECT_NAME, false);
+
+      assertThat(getMeasureAsInt(PROJECT_NAME, "ncloc")).isEqualTo(29);
+      assertThat(getMeasureAsInt(PROJECT_NAME, "bugs")).isEqualTo(1);
+    }
+
     @Test
     public void non_autogenerated_code_reports_bugs() throws Exception {
       final String PROJECT_NAME = "NotAutoGeneratedFiles";
-      analyzeCoverageTestProject(PROJECT_NAME);
+      analyzeCoverageTestProject(PROJECT_NAME, true);
 
       assertThat(getMeasureAsInt(PROJECT_NAME, "ncloc")).isEqualTo(29);
       assertThat(getMeasureAsInt(PROJECT_NAME, "bugs")).isEqualTo(1);
     }
 
-    private void analyzeCoverageTestProject(String projectName, String... keyValues) throws IOException {
+    private void analyzeCoverageTestProject(String projectName, boolean analyzeGenerated) throws IOException {
       Path projectDir = Tests.projectDir(temp, projectName);
       orchestrator.executeBuild(TestUtils.newScanner(projectDir)
         .addArgument("begin")
         .setProjectKey(projectName)
         .setProjectName(projectName)
         .setProjectVersion("1.0")
-        .setProperties(keyValues));
+        .setDebugLogs(true)
+        .setProperty("sonar.cs.analyzeGeneratedCode", "" + analyzeGenerated));
+
+      if (analyzeGenerated) {
+        setAnalyzeGeneratedCodeToTrue(projectDir);
+      }
 
       TestUtils.runMSBuild(orchestrator, projectDir, "/t:Rebuild");
 
       orchestrator.executeBuild(TestUtils.newScanner(projectDir)
         .addArgument("end"));
     }
+
+  private static void setAnalyzeGeneratedCodeToTrue(Path projectDir) throws IOException {
+    // FIXME unfortunately, setting the property inside SQ does not propagate in due time to these files

```

Comment:
S4MSB queries the SQ API to get the settings and creates the 2 XML files. Unfortunately, during the ITs, these files do not contain the up-to-date `sonar.cs.analyzeGeneratedCode` setting, and this is why I am manually changing the file.

When testing manually, this problem does not occur.

Reply:
Let's have a chat about this to see if we can work out what is going on (having a test that we don't understand is a red flag).

From the look of it, I would expect `orchestrator.executeBuild(...).setProperty` to set a command line argument that is passed to the scanner. I'd expect the value for that property to be used by the scanner-cli, as any command line property will should override the server-side property value. I can't see any code seems to be setting the property value on the server.

However, I think there might be a bug in the way the Scanner for MSBuild generates the `SonarLint.xml` file and analysis config file - it may only be using the values from the server, and not overriding them with the value from the command line.

So, my guess (I haven't debugged through) is that what is happening is:
* this test is setting the value on the command line, not on the server.
* this Scanner for MSBuild isn't writing the correct value to either the `SonarLint.xml` file or the analysis config file.
* the code in this method is fixing up both files with the correct value.
* the `sonar-project.properties` file is generated from the analysis config file, so it contains the correct value.
* the scanner-cli API is querying for the setting value, and is correctly overriding the default from the server with the value from the properties file.

Reply:
as discussed with @duncanp-sonar , this problem may be related to SonarSource/sonar-scanner-msbuild#699

Reply:
In the end I decided to use the SQ WS API